subway_prediction_ml/subway_pipeline.py

```diff
@@ -0,0 +1,75 @@
+import pandas as pd
+import os
+from sklearn.compose import ColumnTransformer
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import OneHotEncoder, StandardScaler
+
+from actuals_adder import ActualsAdder
+from destinations_adder import DestinationsAdder
+from locations_adder import LocationsAdder
+from terminal_modes_adder import TerminalModesAdder
+
+
+class SubwayPipeline():
+    def __init__(
+        self,
+        actuals_path=os.path.join("datasets", "pa_datapoints.csv"),
+        locations_path=os.path.join("datasets", "locations.csv"),
+        patterns_path=os.path.join("datasets", "patterns.csv"),
+        terminals_path=os.path.join("datasets", "terminal_datapoints.csv"),
+        vehicles_path=os.path.join("datasets", "vehicle_datapoints.csv")
+    ):
+        self.actuals_path = actuals_path
+        self.locations_path = locations_path
+        self.patterns_path = patterns_path
+        self.terminals_path = terminals_path
+        self.vehicles_path = vehicles_path
+
+    def load(self):
+        vehicle_datapoints = self._load_vehicle_datapoints()
+
+        terminal_modes_adder = TerminalModesAdder(
+            self.locations_path,
+            self.patterns_path,
+            self.terminals_path
+        )
+        locations_adder = LocationsAdder(self.locations_path)
+        destinations_adder = DestinationsAdder(self.locations_path)
+        actuals_adder = ActualsAdder(self.actuals_path)
+        onehot_columns = [
+            'current_location_id',
+            'terminal_gtfs_id',
+            'automatic',
+            'destination_gtfs_id',
+            'event_type'
+        ]
+        column_transformer = ColumnTransformer([
+            ('1hot', OneHotEncoder(), onehot_columns)
+        ], remainder="passthrough")
+
+        pipeline = Pipeline([
+            ('terminal_modes_adder', terminal_modes_adder),
+            ('locations_adder', locations_adder),
+            ('destinations_adder', destinations_adder),
+            ('actuals_adder', actuals_adder),
+            ('col_transformer', column_transformer),
+            ('std_scaler', StandardScaler())
+        ])
+        return pipeline.fit_transform(vehicle_datapoints)
+
+    # Build a dataframe with all logged vehicle datapoints
+    def _load_vehicle_datapoints(self):
+        return pd.read_csv(
```

Comment:
@losvedir we do actually have a `dropna` with no arguments at one point in the `ActualsAdder` that will effectively drop any row containing a null value. I don't actually remember what `offset_departure_seconds_from_now` is, what's the problem with `dropna`'ing that?

Reply:
@losvedir but in any case if we're doing to do an across-the-board `dropna` it would be smart to do so early in the pipeline, to cut as many rows as possible as early as possible.

Reply:
I think `dropna` with no arguments is sort of a sledgehammer to bring to bear on the problem, since there's different ways of handling missing data. I tend to think we shouldn't have a `dropna` with no arguments tucked away in any particular step in the pipeline, and that each step should consider what to do with the data that it personally deals with. Since this is where we load this CSV, I think it makes sense to consider here when to drop data it just loaded, and not reply on `ActualsAdder` down the line (although, it could drop data that's relevant for its purposes, too).

`offset_departure_seconds_from_now` is how many seconds from now, a given trip's return trip is scheduled to depart. The idea is if we're considering a northbound Red Line vehicle at Davis and wanting to predict its arrival at Harvard, we need to consider how long it's scheduled to wait at Alewife (it won't necessarily just turn around as fast as it can go).

However, not all trips have scheduled return trips. This is the case for all Green Line and Mattapan, and also sometimes Red/Orange/Blue whenever we create an "ADDED-" trip for a vehicle, which if you check realtime-dashboard you'll see we do from time to time. (I see 3 "ADDED-" trips on the Red Line right now, out of about 25).

So, `offset_departure_seconds_from_now` is definitely useful if we have it, but if we don't I'm not totally certain how to handle it. Ideally we'd interpolate in a value that means "don't wait" but that sounds complex to do. We could potentially do it on the RTR side when logging the value, now that I think about it, or else maybe bin the time.

Reply:
I don't get what you mean by a value that means "don't wait", or binning times, could you expand on those?

Reply:
For "don't wait", I meant, filling in a reasonable non-zero `offset_departure_seconds_from_now` which doesn't have the train waiting for some scheduled departure. It's hard, though, because a NB train at Davis might have an `o_d_s_f_n` for its return trip of just a couple minutes, while one down at Broadway may be 20+.

For binning, I mean essentially one-hotting the column with an additional "is scheduled", so instead of:

```
offset_departure_seconds_from_now
------------------------------------
40
null
80
```

we have:

```
scheduled?, scheduled_0-60, scheduled_61-120,
---------------------------------------------
1,1,0
0,0,0
1,0,1
```

Reply:
@losvedir I don't think coming up with a reasonable nonzero value is practicable, so I'll add another pipeline stage to do some binning.